lib/logflare/context_cache.ex

```diff
@@ -32,43 +41,52 @@ defmodule Logflare.ContextCache do
         value
 
       {:ok, {:cached, value}} ->
-        index_keys(context, cache_key, value)
+        # already cached don't re-index
         value
     end
   end
 
+  @doc """
+  This function is called from the CacheBuster process when a new record comes in from the Postgres
+  write-ahead log. The WAL contains records. From those records the CacheBuster picks out
+  primary keys.
+
+  The records ARE the keys in the reverse cache (the ContextCache).
+
+  We must:
+   - Find the key by the record primary key
+   - Delete the reverse cache entry
+   - Delete the cache entry for that context cache e.g. `Logflare.Users.Cache`
+  """
+
   def bust_keys(values) do
-    {:ok, keys} = Cachex.keys(@cache)
-
-    total =
-      Enum.count(keys, fn {token, cache_key} = key ->
-        with true <- token in values,
-             {context, _} = token,
-             context_cache = cache_name(context),
-             {:ok, true} <- Cachex.del(context_cache, cache_key) do
-          Cachex.del(@cache, key)
-        end
-
-        true
+    for {context, primary_key} <- values do
+      filter = {:==, {:element, 1, :key}, {{context, primary_key}}}
+      query = Cachex.Query.create(filter, {:key, :value})
+      context_cache = cache_name(context)
+
+      Logflare.ContextCache
+      |> Cachex.stream!(query)
+      |> Enum.each(fn {k, v} ->
+        Cachex.del(context_cache, v)
+        Cachex.del(@cache, k)
       end)
-
-    :telemetry.execute(
-      [:logflare, :context_cache, :busted],
-      %{count: total},
-      %{}
-    )
+    end
```

Comment:
I wonder if we maybe can do everything in one step. So create match spec list for all possible matched values, and then evict them in single ETS query. I wonder if that would be faster rather than doing separate queries. I would need to benchmark that.

Reply:
I think we could but right now the cache for the context (`Logflare.Users.Cache`) is a different ETS table than the reverse index (`Logflare.ContextCache`) so we'd need to refactor to use one ETS table for all of them (and probably partition them).